### DISCUSSION TITLE
Signing key with TortoiseGit and GitHub

### DISCUSSION THREAD

**Discussion by LudiusMaximus** [community/community#23205] [New to GitHub]
<p>When I commit/push to my projects using TortoiseGit and my GitHub credentials, the commits are not attributed to my GitHub profile. I guess this is because I have not set up a signing key with TortoiseGit? (I assumed that my GitHub username and password would suffice to identify me but apparently this ist not the case…)</p>
<p>So I wanted to set up a key but I got confused.</p>
<p>Under TortoiseGit you can setup a Putty Key (.ppk) for each repository:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/d95eb5dbec7fb5dce18e0e38909bb49cb8e7b59b.jpeg" data-download-href="/uploads/short-url/v0WwGnZcGbLkcd2k1uzubZwLKGf.jpg?dl=1" title="Capture.JPG" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/4678285/181075590-72fbda46-ee05-45d7-af9d-e34d4d54f522.jpeg" alt="Capture.JPG" data-base62-sha1="v0WwGnZcGbLkcd2k1uzubZwLKGf" width="690" height="486" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/d95eb5dbec7fb5dce18e0e38909bb49cb8e7b59b_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Capture.JPG</span><span class="informations">831×586 60.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>But when I create a keypair with puttygen.exe,</p>
<p>the public key starts with</p>
<p>---- BEGIN SSH2 PUBLIC KEY ----<br>
Comment: “rsa-key-20180719”<br>
AAAAB3NzaC1yc2EAAA…</p>
<p>and github will not accept this when I try to add a new SSH key:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/42db5521b5f1bfde0df70b2d362d80055b315767.jpeg" data-download-href="/uploads/short-url/9xrqIel8ZUNX8rPHmDprfFf0Umz.jpg?dl=1" title="Capture2.JPG" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/4678285/181075591-00d7ce58-6f7e-45da-858d-f548ae9eaaa0.jpg" alt="Capture2.JPG" data-base62-sha1="9xrqIel8ZUNX8rPHmDprfFf0Umz" width="690" height="268" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/42db5521b5f1bfde0df70b2d362d80055b315767_2_690x268.jpg, https://aws1.discourse-cdn.com/github/optimized/1X/42db5521b5f1bfde0df70b2d362d80055b315767_2_1035x402.jpg 1.5x, https://aws1.discourse-cdn.com/github/original/1X/42db5521b5f1bfde0df70b2d362d80055b315767.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/42db5521b5f1bfde0df70b2d362d80055b315767_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Capture2.JPG</span><span class="informations">1295×504 56.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>On the other hand, when I generate a key pair as described on the GitHub Page with</p>
<pre><code>ssh-keygen -t rsa -b 4096 -C "_your\_email@example.com_"
</code></pre>
<p>I get a key pair, whose public key is accepted by GitHub, but the private key is not a .pkk file,</p>
<p>so I guess it would not be accepted by TortoiseGit as a “Putty Key” (did not try)?</p>
<p>How is this done the right way?</p>
<p>Thanks.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>A signing key won’t help associate commits with your GitHub account. What you need to do is <a href="https://tortoisegit.org/docs/tortoisegit/tgit-dug-settings.html#tgit-dug-settings-config" rel="nofollow noopener">set your user name and email address</a>, and the email address must match one of the email addresses that you have configured for your GitHub account. See the <a href="https://help.github.com/articles/about-commit-email-addresses/" rel="nofollow noopener">GitHub help documentation</a> and the Support Protips article <a href="https://github.community/t/why-is-my-commit-associated-with-the-wrong-person/10199">Why is my commit associated with the wrong person?</a> for further information.</p>

**Comment by NoxFfredrika** [comment]
<p>I had this same problem with the SSH keys. The original post is asking about commit attribution and the solution posted seems to address that.</p>
<p>However, I’d like to know if there is a solution to the SSH key problem itself. Is there a way to use SSH with TortoiseGIT and GitHub without having these key conflicts?</p>
<p>Thanks much for any advice anyone has!</p>

**Comment by Nainoia-Inc-Admin** [comment]
<p>Related question when trying to push to Github with TortioseGit.  I have copied my public SSH into my Github account.  And within TortioseGit I point to my SSH key.  However, what is the Github URL supposed to be???  Please provide an example.</p>
<p>This doesn’t work…</p>
<p>ssh://git@github.com:&gt;/Repository.git</p>

**Comment by airtower-luna** [comment]
<p>The scheme is <code>git@github.com:user/repo.git</code>. You can just click on the green “Code” button on your repository page and select “SSH” in the popup to get the exact URL.</p>

**Comment by blounsbury-usbr** [comment]
I'm sure this is a no-no to resurrect an old thread, but I'm having essentially the same issue. Commits do not show up as Verified on Github when committing using TortoiseGit. If I commit from the command line it works.

My username and email is set correctly in Github and TortoiseGit.

Viewing the effective config from TortoiseGit everything seems fine. I have a user.signingkey and it is set to ssh.

![image](https://github.com/community/community/assets/1508213/186b1165-ba9d-4988-a7aa-5669a643648c)

Like I mentioned from the command line it works just fine.

![image](https://github.com/community/community/assets/1508213/4a8fcd1f-fbae-45bd-903d-1ae60ad598d4)

Just can't figure out what needs set from the TortoiseGit side to get this working.

**Comment by jbostrus** [comment]
So I kind of feel bad to Necro this discussion again. But seeing as I didn't actually see a functional resolution listed yet, and I just came across this page, and I did finally get it working. I thought I would share.

So from the original answer, you would need the email and such in Tortoise to match GitHub. From there, the information from @blounsbury-usbr got me going in the right direction. Ultimately [this GitHub page](https://docs.github.com/en/authentication/managing-commit-signature-verification/telling-git-about-your-signing-key) detailing setting up the client from the command line was necessary.

@LudiusMaximus original post noted that Puttygen produced a public file that wasn't suitable for GitHub. In Puttygen the UI has the rsa public key right there to copy/paste.
![image](https://github.com/community/community/assets/12855515/19e4c815-d0ed-46eb-86c5-5d8fb9873f6c)
(dummy generated key just for the screenshot - not an actively used key)

In the post if using a Putty key in the UI there, it also would need to be set to use the GitHub SSH address not the https URL for the remote.

For the rest of this I personally ended up just doing it all from the command line because I couldn't find all the relevant settings in the TortoiseGit UI that easily. Afterwards I realized that the file path for the SSH public key can just be pasted into the `Signing Key ID` field on the Git settings page in the TortoiseGit UI. The other 2 settings listed below I couldn't find UI entries for. So they needed the command line or direct `.gitconfig` edits anyway.

1. First in GitHub add your public SSH key for signing. If you already have it for connection, you'll need to add it again and select signing. So it will be listed twice. 
![image](https://github.com/community/community/assets/12855515/0cd53cf5-fd36-4ee7-8f67-971d1febc387)
![image](https://github.com/community/community/assets/12855515/0f4cc6cd-a2a9-4ead-b15c-7deb3874a442)

2. From a command line run 
  - git config --global gpg.format ssh
  - git config --global user.signingkey /PATH/TO/.SSH/KEY.PUB
  This last one is what tells it to actually sign things.
  - git config --global commit.gpgsign true

If you'd rather edit the Global .gitconfig directly you could do that too. The relevant changes to make would be

Under `[user]` add
```ini
	signingkey = /PATH/TO/.SSH/KEY.PUB
```
If you're on Windows replace any `/` with 2 `\\` ie `C:\\PATH\\TO\\.SSH\\KEY.PUB`

Then add the following if they don't already exist
```ini
[gpg]
	format = ssh
[commit]
	gpgsign = true
```
so it would look something like this
![image](https://github.com/community/community/assets/12855515/3d7b3962-bf74-45ca-8798-b47c901e4e6c)

And before I get spammed about how insecure it is to keep the key there, this is only to illustrate how it would look. That's not where I keep my keys.

If all was done correctly then after the next local commit and then push to GitHub you should have a nice green Verified on the commit.

**Reply by robross0606** [reply]
These steps do not appear to fully work in Windows.  It keeps telling me the file doesn't exist at the provided path even when it does.
